text/000-symbol-sets.md

```diff
@@ -0,0 +1,80 @@
+- Title: Symbol Sets
+
+- Drivers: Michael Soegtrop
+
+----
+
+# Summary
+
+In automation the most reliable and fastest way to reduce terms is frequently to use the `lazy` reduction with an explicit list of symbols to expand, that is:
+```
+lazy [ <some symbols> ].
+```
+
+The problem with this is that the symbol sets can be lengthy - a few 100 is not rare in practice - and difficult to maintain.
+Typically the symbol sets are all symbols from specific modules with individual inclusions / exclusions.
+An effective solution would be to be able to declare named symbol sets using
+- individual symbols
+- all symbols in a module without recursion into sub modules
+- all symbols in a module with recursion into sub modules
+- set operations union, intersection and difference
+
+# Motivation
+
+In proof automation one frequently has to reduce terms which contain user supplied terms, that is terms whose content the automation does not restrict.
+Applying `simpl` or `cbn` to such terms does not work reliably, because there are examples for term where `simpl` and `cbn` take a very long time.
+If the user supplied terms are of this nature, the overall tactic will also take a long time, even though the task at hand in the automation would be trivial.
+Besides in automation one usually does not want the user supplied terms to be changed at all.
+
+What is done in some Coq developments is to hide the user supplied terms behind opaque definitions and then use full evaluation.
+This works, but the hiding and unhiding of user terms can take substantially longer than the reduction itself.
+
+The only really good option is using `lazy` or `cbv` with a delta symbol set, which just contains private definitions of the automated tactic itself.
+This  requires an effective way to manage symbol sets.
+
+Note that this method frequently involves copying some basic functions like list append.
+This is usually not a problem - it can become a problem if the automation needs e.g. arithmetic.
+It would make sense to have a mechanism to copy the contents of a module under a new logical path, so that one can add the copy to the delta reduction list, but leave the original symbols unreduced.
+But this is an independent topic and should be discussed in a separate CEP.
+
+Note that this related to [Declare Reduction](https://coq.inria.fr/doc/V8.19.0/refman/proofs/writing-proofs/equality.html#coq:cmd.Declare-Reduction).
+
+# Detailed design
+
+The syntax doesn't really matter a lot - it is fine to choose something which is easy to implement.
+
+A possible syntax would be:
+- Command **Declare Symbol Set** *ident* **:=** *symbol_set_expr*
+- *symbol_set_expr* ::= **Module** *qualid* | **Module Recursive** *qualid* | *reference* | *symbol_set_expr* [**+ * -**] *symbol_set_expr* | **(** *symbol_set_expr* **)**
```

Comment:
Indeed the syntax is net very well thought through - mostly because I know little about how such "global syntax" messes with the parser and notation system. IMHO designing good Coq syntax is an art on its own, which I would prefer to leave to those who understand it.